### ISSUE TITLE
On Windows, results depend on file lineends used by Git

### ISSUE THREAD

**Issue by Blaisorblade** [nikita-volkov/neat-interpolation#14]
Values written using neat-interpolation use the newlines of the source file. On Windows that means they'll likely contain `\r\n` instead of `\n`. That's unfortunate and unlikely to be intended, so I'd guess neat-interpolation should canonicalize newlines to `\n`.

Rationale 1: this package is used for string constants, so they better be constant.
Rationale 2: on Windows the actual newlines in a Git-managed project depend on how exactly Git was setup. The setup recommended by Git for Windows converts 2.9.2 them to Windows newlines.
Rationale 3: If I didn't use raw strings, I'd be writing `\n` in those strings. That's because the conversion to `\r\n` is done at the lowest possible level (when writing to files opened in text mode).

Use case:
A colleague used `neat-interpolation` (which is cool) in a test case I'm porting to Windows. However, I got the following:

```
Failures:

  src/test\Stack\GhciSpec.hs:54:
  1) Stack.Ghci.GHCi, Script rendering, should render GHCi scripts, with one lib                                                                      rary package
       expected: ":add Lib.A\r\n:module + Lib.A\n"
        but got: ":add Lib.A\n:module + Lib.A\n"
```

The expected value is written with a `neat-interpolation` quasiquote, and it has a `\r\n` instead of `\n`. To confirm my analysis above, I 
https://ci.appveyor.com/project/snoyberg/stack/build/1.0.2076

I didn't find any explicit mention of newlines in the package, but I could confirm this happens because even on Windows (somewhat sensibly) `lines` only processes `\n`, leaving `\r` alone:
```
> lines "a\r\nb"
["a\r","b"]
```

**Comment by nikita-volkov** [comment]
I don't quite understand the concern here. Am I correct that the preprocessor retains the same line -separators as were used in the source-code sent to the compiler?

**Comment by Blaisorblade** [comment]
Yes, you're correct.
I'm trying to explain why in many use cases that's not what you want. Or actually, I believe that's never what you want, but I anticipate that might be debatable.

**Comment by Blaisorblade** [comment]
A better explanation. Do you agree that I should be able to replace `"a\nb\n"` with
```
[text|
a
b
|]
``` and get the same result, irrespective of the used newlines in the source file?

If so, your quasiquoter should *modify* newlines to be `\n`.

**Comment by nikita-volkov** [comment]
No problem. I completely disagree however.

I think you're projecting on this library the requirements, which one would expect from the editor (i.e., the automatic line-separator normalisation). All the imaginable code-editors out there have the related settings. 

The library does only what it claims to do. If the user's code contains the `\r\n` character sequences it might just as well be by intent and I see no other unambiguous way to approach that. But, above all, I see no need in that, because, again, it's not of any library's concern to guess, which line endings the user intends to have.

**Comment by Blaisorblade** [comment]
> But, above all, I see no need in that, because, again, it's not of any library's concern to guess, which line endings the user intends to have.

My proposal would not pick Unix line endings for users; Windows line endings are converted to `\n` when opening a file in text mode, in all IO libraries around including Haskell (http://hackage.haskell.org/package/base-4.9.0.0/docs/GHC-IO-Handle-FD.html#v:openFile). This preserves the invariant that text strings in a program always use `\n`. Which is why putStr "A\nB\n" will use native endings on all platforms, and why lines and unlines have the same implementation on all platforms.

The truth is, this is probably a GHC bug.

Binary files are a separate matter, but that's not your usecase.

> I think you're projecting on this library the requirements, which one would expect from the editor (i.e., the automatic line-separator normalisation). All the imaginable code-editors out there have the related settings.

Nope. I can't setup the editors of people checking out my sources, but by default (I guess) they'll preserve existing newlines. But again, Git won't.

# Prior art

Other interpolation libraries work as I describe. Without even necessarily documenting it.

interpolatedstring-qq in Haskell:
http://hackage.haskell.org/package/interpolatedstring-qq-0.2/docs/src/Text-InterpolatedString-QQ.html#istr

Perl6 once violated the convention I describe, intentionally. Their entire testsuite became sensitive to users Git settings, and they changed their mind in 5 days after a ticket was filed.
https://rt.perl.org/Public/Bug/Display.html?id=126881
https://www.nu42.com/2015/12/perl6-newline-translation-broken.html
https://www.reddit.com/r/perl6/comments/3we6d6/newline_translation_in_perl6_is_broken/

**Comment by Blaisorblade** [comment]
Here's a specific issue reaching the same conclusion:
https://github.com/23Skidoo/raw-strings-qq/issues/1

**Comment by nikita-volkov** [comment]
Hmm. I'm honestly conflicted on this. On the one hand, you make arguments which I acknowledge and you're obviously informed on the issue's spread, on the other, this is a very ambiguous case.

I think the most "politically correct" way to approach this would be to introduce another quasi-quoter, which would provide the behaviour you expect. I however don't have time to approach this myself, I'm open to accept such a pull-request though.

**Comment by Blaisorblade** [comment]
Thanks, I'll take a look at making a PR!

**Comment by sellout** [comment]
**tl;dr**: I agree that a separate `\r`-removing `QuasiQuoter` is the best approach, even though I didn’t when I started writing this comment. I also think a note in the Haddock with a brief explanation and link to the GHC issue would be very helpful.

---

Line endings are complicated, but the [Haskell (or at least GHC?) documentation on this](https://hackage.haskell.org/package/base-4.21.0.0/docs/System-IO.html#g:24) is pretty clear (if not particularly discoverable).

This is actually [a bug in GHC](https://gitlab.haskell.org/ghc/ghc/-/issues/14876), since as the docs say, “In Haskell, a newline is always represented by the character '\n'.” Your quasi-quoting function is given a `String`, which means that it should already have had that line-ending conversion applied to it. Unfortunately, the bug has been open for seven years.

I think you’re right that correct handling would be tricky – you don’t want to blindly remove `\r`, because if a user does explicitly put a `\r` into the content
``` haskell
let cr = "\r" in [trimming|$cr|]
```
you want to preserve that. And replacing `\r\n` with `\n` can break on non-Windows systems as it would cause
``` haskell
let cr = "\r"
 in [trimming|$cr
|]
```
(`\r\n`) to become equivalent to
``` haskell
[trimming|
|]
```
(`\n`). And, very unfortunately, even doing it conditionally based on the platform won’t work, because
1.  Windows is content to have files with `\n` line endings these days (giving those files the same edge cases as non-Windows), and even if it weren’t,
2. GHC itself in inconsistent (as mentioned in [an issue comment](https://gitlab.haskell.org/ghc/ghc/-/issues/14876#note_150687)) – if I write `trimming_file = quoteFile trimming`, then `[trimming_file|my-file.yaml|]` (on Windows) will have `\n` even if my-file.yaml contains `\r\n` on disk.

So, I don’t know there’s anything you can do in Haskell code to work around this properly – you would need access to the Haskell source file containing the quasi-quoted text to be able to determine its line-endings. So the best option is probably a new `QuasiQuoter` with a better name than `[trimming_andIPromiseIDidNotPutAnyCRsInHere|...|]` that replaces `\r\n` with `\n`. I would suggest that this version should actually be the “default” behavior[^1], since it’s almost always what people want, but that would technically be a breaking change.

[^1]: With the Haddock explaining that there’s a variant if you really need to interpolate a literal `\r` before a `\n` for some reason.

**Comment by sellout** [comment]
Ugh, I just realized something after looking at @Blaisorblade’s [commit](https://github.com/Blaisorblade/neat-interpolation/commit/a49dede8fc0e34e27b0c4cad5d8251cded58f59b) (that never became a PR for some reason).

If the `\r\n` -> `\n` conversion is done before interpolation, it should always do the right thing. Because I don’t think there’s any way to type a `\n` directly into the quasi-quoted text, right? It has to be done via interpolation, as in my comment above.

So, I’m back to thinking that while, yes, GHC should fix this, doing the replacement on the input string will always do the right thing. And, in fact, even a simple `\r` removal (slightly simpler than replacement) will also do the right thing. So, the existing `QuasiQuoter`s should be changed, and no new ones are needed.
